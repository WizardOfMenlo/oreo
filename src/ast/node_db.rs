//! Something

use super::untyped::Node;
use std::collections::HashMap;
use std::marker::PhantomPinned;
use std::pin::Pin;
use std::ptr::NonNull;

/// An id for a node
#[derive(Debug, Clone, Copy, Hash, Eq, PartialEq, Ord, PartialOrd)]
pub struct NodeId(usize);

/// A node that is flat
#[derive(Debug, Clone)]
pub struct IdNode {
    childrens: Vec<NodeId>,
}

impl IdNode {
    /// Gets the children of this node
    pub fn children(&self) -> &[NodeId] {
        &self.childrens
    }
}

/// A database that efficiently allows for querying and retrivieng nodes
#[derive(Debug)]
pub struct NodeDb<'a> {
    node: Node<'a>,
    id_nodes: HashMap<NodeId, IdNode>,
    nodes: HashMap<NodeId, NonNull<Node<'a>>>,
    start_node_id: NodeId,
    _phantom: PhantomPinned,
}

impl<'a> NodeDb<'a> {
    /// Creates a db from a node
    pub fn new(node: Node<'a>) -> Pin<Box<Self>> {
        // Ok this is complex and merit some discussion
        // In order to efficiently (aka not storing nodes and children twice)
        // store nodes, we aim to store a reference to the children of a stored not
        // Ordinarly, rust wouldn't like that, so we need to bend the rules a little
        // First of all, this is a self referential struct, i.e. one that has references upon itself
        // (in particular node is borrowed by nodes)
        // Since moving the struct would invalidate this references, we use the Pin API, which allows to
        // to define types that cannot be moved
        // Using this, we can make our thing work and be efficient!
        let db = NodeDb {
            node,
            _phantom: PhantomPinned,
            // Temp
            id_nodes: HashMap::new(),
            nodes: HashMap::new(),
            start_node_id: NodeId(0),
        };

        let mut boxed = Box::pin(db);

        let (id, id_nodes, nodes) =
            add_node(&boxed.node, NodeId(0), HashMap::new(), HashMap::new());

        // I claim this is safe, as we are never moving the box and such not
        // violating the Pin API
        unsafe {
            let reference = Pin::as_mut(&mut boxed);
            let inner = Pin::get_unchecked_mut(reference);
            inner.start_node_id = NodeId(id.0.saturating_sub(1));
            inner.id_nodes = id_nodes;
            inner.nodes = nodes;
        }

        boxed
    }

    /// All the ids that this struct owns
    pub fn get_all_node_ids<'b>(&'b self) -> impl Iterator<Item = NodeId> + 'b {
        self.id_nodes.keys().cloned()
    }

    /// Get a flat representation of the node
    pub fn get_flat_node(&self, id: NodeId) -> Option<&IdNode> {
        self.id_nodes.get(&id)
    }

    /// Get the start id (ie the top node id)
    pub fn start_id(&self) -> NodeId {
        self.start_node_id
    }

    /// Get a reference to the full node
    pub fn get_node(&self, id: NodeId) -> Option<&Node<'a>> {
        // Note this should be safe as p always refer to a node in the
        // struct
        self.nodes.get(&id).map(|p| unsafe { p.as_ref() })
    }

    /// Get all the children that this node has
    pub fn all_children(&self, id: NodeId) -> Vec<NodeId> {
        let mut res = Vec::new();
        for child in self
            .get_flat_node(id)
            .map(|n| n.children())
            .unwrap_or_default()
        {
            res.push(*child);
            res.extend(self.all_children(*child));
        }
        res
    }
}

type NodeIdMap = HashMap<NodeId, IdNode>;
type NodeMap<'a> = HashMap<NodeId, NonNull<Node<'a>>>;

// Not in class to circumvent (safely) rust borrowing rules
fn add_node<'a>(
    n: &Node<'a>,
    mut id: NodeId,
    mut id_nodes: NodeIdMap,
    mut nodes: NodeMap<'a>,
) -> (NodeId, NodeIdMap, NodeMap<'a>) {
    let mut ids = Vec::new();
    for c in n.children() {
        let (n_id, n_id_nodes, n_nodes) = add_node(c, id, id_nodes.clone(), nodes.clone());
        id = n_id;
        id_nodes.extend(n_id_nodes);
        nodes.extend(n_nodes);
        ids.push(NodeId(id.0 - 1));
    }

    id_nodes.insert(id, IdNode { childrens: ids });

    nodes.insert(id, NonNull::from(n));

    (NodeId(id.0 + 1), id_nodes, nodes)
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_debug_snapshot;

    fn node_from_str(input: &str) -> Node {
        use crate::lexer::lexicalize;
        use crate::lexer::scanner::scan;
        use crate::parser::parse;

        parse(lexicalize(scan(input)))
    }

    // We use this to test, as the Db does not serialize consistently
    fn get_all_nodes<'a>(db: &NodeDb<'a>) -> Vec<Node<'a>> {
        let mut ids: Vec<_> = db.get_all_node_ids().collect();
        ids.sort_by(|a, b| a.0.cmp(&b.0));

        ids.into_iter()
            .map(|i| db.get_node(i).expect("This should be failproof"))
            .cloned()
            .collect()
    }

    #[test]
    fn node_db_minimal() {
        let input = "program fib begin var n; end";
        let db = NodeDb::new(node_from_str(input));
        let nodes = get_all_nodes(&db);
        assert_debug_snapshot!(nodes);
    }

    #[test]
    fn node_db_minimal_assignments() {
        let input = "program fib begin var n := \"Hello\"; var m := 2; var s := (2); end";
        let db = NodeDb::new(node_from_str(input));
        let nodes = get_all_nodes(&db);
        assert_debug_snapshot!(nodes);
    }

    #[test]
    fn node_db_print_simple() {
        let input = "program fib begin print \"Hello\\n\"; println 2; get x; end";
        let db = NodeDb::new(node_from_str(input));
        let nodes = get_all_nodes(&db);
        assert_debug_snapshot!(nodes);
    }
}
