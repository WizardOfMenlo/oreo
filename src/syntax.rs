pub struct Identifier<'a>(&'a str);

pub struct Program<'a> {
    id: Identifier<'a>,
    compund: Compound<'a>,
}

pub struct Compound<'a> {
    statements: Vec<Statement<'a>>,
}

pub enum Statement<'a> {
    Decl(Decl<'a>),
    Print(Print<'a>),
    While(While<'a>),
    If(If<'a>),
    Assign(Assign<'a>),
}

pub struct Decl<'a> {
    id: Identifier<'a>,
    expr: Option<Expr<'a>>,
}

pub enum Print<'a> {
    Print(Expr<'a>),
    Println(Expr<'a>),
    Get(Identifier<'a>),
}

pub struct While<'a> {
    condition: Bool<'a>,
    compound: Compound<'a>,
}

pub struct If<'a> {
    condition: Bool<'a>,
    if_branch: Compound<'a>,
    else_branch: Option<Compound<'a>>,
}

pub struct Assign<'a> {
    id: Identifier<'a>,
    expr: Expr<'a>,
}

pub enum Expr<'a> {
    Addition(Box<BinaryOperator<'a>>),
    Subtraction(Box<BinaryOperator<'a>>),
    Multiplication(Box<BinaryOperator<'a>>),
    Division(Box<BinaryOperator<'a>>),
    Bool(Box<Bool<'a>>),
    Expr(Box<Expr<'a>>),
    Literal(Literal<'a>),
    Identifier(Identifier<'a>)
}

pub enum Literal<'a> {
    Str(&'a str),
    Int(isize)
}

pub struct BinaryOperator<'a> {
    left: Expr<'a>,
    right: Expr<'a>,
}

pub enum Bool<'a> {
    LessThan(BooleanRelOperator<'a>),
    GreaterThan(BooleanRelOperator<'a>),
    LesserOrEquals(BooleanRelOperator<'a>),
    GreaterOrEquals(BooleanRelOperator<'a>),
    Equals(BooleanRelOperator<'a>),
    And(Box<BooleanBinaryOperator<'a>>),
    Or(Box<BooleanBinaryOperator<'a>>),
    Not(Box<Bool<'a>>),
    BooleanLiteral(bool),
}

pub struct BooleanRelOperator<'a> {
    first: Expr<'a>,
    second: Expr<'a>,
}

pub struct BooleanBinaryOperator<'a> {
    first: Bool<'a>,
    second: Bool<'a>,
}