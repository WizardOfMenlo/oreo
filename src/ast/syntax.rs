//! Typed syntax that is backed by Node

use super::node_db::{NodeDb, NodeId};
use super::untyped::{Node, NodeType};

/// Macro for simple definition of a syntax node
macro_rules! syntax_node {
    ($id:ident) => {
        /// A syntax node for an $id
        #[derive(Debug, Clone, Copy, Hash, Eq, PartialEq)]
        pub struct $id(NodeId);

        impl $id {
            /// Instantiate the syntax node
            pub fn new(node: NodeId) -> Self {
                $id(node)
            }

            /// Get node id
            pub fn get_id(self) -> NodeId {
                self.0
            }

            /// Get the raw node
            pub fn get_node<'a, 'b>(self, db: &'b NodeDb<'a>) -> &'b Node<'a> {
                db.get_node(self.0).expect("Invalid syntax node")
            }

            /// Get children
            pub fn children<'a, 'b>(self, db: &'b NodeDb<'a>) -> &'b [NodeId] {
                db.get_flat_node(self.0)
                    .expect("Invalid syntax node")
                    .children()
            }
        }
    };
}

syntax_node!(Program);

impl Program {
    /// Get the program name
    pub fn id<'a>(self, db: &NodeDb<'a>) -> Identifier {
        Identifier(self.children(db)[0])
    }

    /// Program main body
    pub fn compound<'a>(self, db: &NodeDb<'a>) -> Compound {
        Compound(self.children(db)[1])
    }
}

syntax_node!(Compound);

impl Compound {
    /// Get the inner statements
    pub fn statements<'a>(self, db: &NodeDb<'a>) -> Vec<Statement> {
        self.children(db).iter().map(|n| Statement(*n)).collect()
    }
}

syntax_node!(Statement);

impl Statement {
    /// Convert the statement to a more specific one
    pub fn downcast<'a>(self, db: &NodeDb<'a>) -> StatementType {
        let n = self.0;
        match self.get_node(db).ty() {
            NodeType::Decl => StatementType::Decl(Decl(n)),
            NodeType::PrintStat => StatementType::PrintStat(PrintStat(n)),
            NodeType::While => StatementType::While(While(n)),
            NodeType::If => StatementType::If(If(n)),
            NodeType::Assign => StatementType::Assign(Assign(n)),
            NodeType::FunctionCall => StatementType::FunctionCall(FunctionCall(n)),
            NodeType::FunctionDecl => StatementType::FunctionDecl(FunctionDecl(n)),
            NodeType::Return => StatementType::Return(Return(n)),
            _ => panic!("Invalid node constructed"),
        }
    }
}

#[allow(missing_docs)]
#[derive(Debug, Clone)]
/// The possible types that a statement can be
pub enum StatementType {
    Decl(Decl),
    PrintStat(PrintStat),
    While(While),
    If(If),
    Assign(Assign),
    FunctionCall(FunctionCall),
    FunctionDecl(FunctionDecl),
    Return(Return),
}

syntax_node!(Decl);

impl Decl {
    /// The identifier being declared
    pub fn id<'a>(self, db: &NodeDb<'a>) -> Identifier {
        Identifier(self.children(db)[0])
    }

    /// The expression to assign (if any)
    pub fn expr<'a>(self, db: &NodeDb<'a>) -> Option<Expr> {
        self.children(db).get(1).map(|n| Expr(*n))
    }
}

syntax_node!(PrintStat);

impl PrintStat {
    /// Convert to the more specific print typ
    pub fn downcast<'a>(self, db: &NodeDb<'a>) -> PrintTypes {
        let node = self.children(db)[0];
        match self.get_node(db).ty() {
            NodeType::Print => PrintTypes::Print(Print(node)),
            NodeType::Println => PrintTypes::Println(Println(node)),
            NodeType::Get => PrintTypes::Get(Get(node)),
            _ => panic!("Invalid node constructed"),
        }
    }
}

/// Possible print types
#[derive(Debug, Clone)]
#[allow(missing_docs)]
pub enum PrintTypes {
    Print(Print),
    Println(Println),
    Get(Get),
}

syntax_node!(Print);

impl Print {
    /// The expression to print
    pub fn expr<'a>(self, db: &NodeDb<'a>) -> Expr {
        Expr(self.children(db)[0])
    }
}

syntax_node!(Println);

impl Println {
    /// The expression to print
    pub fn expr<'a>(self, db: &NodeDb<'a>) -> Expr {
        Expr(self.children(db)[0])
    }
}

syntax_node!(Get);

impl Get {
    /// The identifier to set
    pub fn id<'a>(self, db: &NodeDb<'a>) -> Identifier {
        Identifier(self.children(db)[0])
    }
}

syntax_node!(While);

impl While {
    /// The condition to evaluate
    pub fn condition<'a>(self, db: &NodeDb<'a>) -> Expr {
        Expr(self.children(db)[0])
    }

    /// The body to be executed
    pub fn compound<'a>(self, db: &NodeDb<'a>) -> Compound {
        Compound(self.children(db)[1])
    }
}

syntax_node!(If);

impl If {
    /// The condition to evaluate
    pub fn condition<'a>(self, db: &NodeDb<'a>) -> Expr {
        Expr(self.children(db)[0])
    }

    /// If true
    pub fn if_branch<'a>(self, db: &NodeDb<'a>) -> Compound {
        Compound(self.children(db)[1])
    }

    /// If false (optional)
    pub fn else_branch<'a>(self, db: &NodeDb<'a>) -> Option<Compound> {
        self.children(db).get(2).map(|n| Compound(*n))
    }
}

syntax_node!(Assign);

impl Assign {
    /// The identifier to set
    pub fn id<'a>(self, db: &NodeDb<'a>) -> Identifier {
        Identifier(self.children(db)[0])
    }

    /// The expression to set it to
    pub fn expr<'a>(self, db: &NodeDb<'a>) -> Expr {
        Expr(self.children(db)[1])
    }
}

syntax_node!(FunctionDecl);

impl FunctionDecl {
    /// Function to declare
    pub fn id<'a>(self, db: &NodeDb<'a>) -> Identifier {
        Identifier(self.children(db)[0])
    }

    /// The arguments
    pub fn args<'a>(self, db: &NodeDb<'a>) -> Vec<FunctionDeclArgs> {
        let children = self.children(db);
        children[1..children.len() - 1]
            .iter()
            .map(|n| FunctionDeclArgs(*n))
            .collect()
    }

    /// The body to be executed
    pub fn compound<'a>(self, db: &NodeDb<'a>) -> Compound {
        let children = self.children(db);
        Compound(children[children.len() - 1])
    }
}

syntax_node!(FunctionDeclArgs);

impl FunctionDeclArgs {
    /// The arg id
    pub fn id<'a>(self, db: &NodeDb<'a>) -> Identifier {
        Identifier(self.children(db)[0])
    }
}

syntax_node!(FunctionCallArgs);

impl FunctionCallArgs {
    /// The expression in the argument
    pub fn expr<'a>(self, db: &NodeDb<'a>) -> Expr {
        Expr(self.children(db)[0])
    }
}

syntax_node!(Return);

impl Return {
    /// The expression to return
    pub fn expr<'a>(self, db: &NodeDb<'a>) -> Expr {
        Expr(self.children(db)[0])
    }
}

syntax_node!(FunctionCall);

impl FunctionCall {
    /// The function id to call
    pub fn id<'a>(self, db: &NodeDb<'a>) -> Identifier {
        Identifier(self.children(db)[0])
    }

    /// The arguments of the call
    pub fn args<'a>(self, db: &NodeDb<'a>) -> Vec<FunctionCallArgs> {
        self.children(db)[1..]
            .iter()
            .map(|n| FunctionCallArgs(*n))
            .collect()
    }
}

syntax_node!(Expr);

impl Expr {
    /// The head of the expression
    pub fn head<'a>(self, db: &NodeDb<'a>) -> Term {
        Term(self.children(db)[0])
    }

    /// Tail of the expression
    pub fn tail<'a>(self, db: &NodeDb<'a>) -> Option<ExprPrime> {
        self.children(db).get(1).map(|e| ExprPrime(*e))
    }
}

syntax_node!(ExprPrime);

impl ExprPrime {
    /// The operation
    pub fn op<'a>(self, db: &NodeDb<'a>) -> BooleanOp {
        self.get_node(db).ty().clone().into()
    }

    /// The right operand
    pub fn rhs<'a>(self, db: &NodeDb<'a>) -> Term {
        Term(self.children(db)[0])
    }

    /// Possible tails
    pub fn tail<'a>(self, db: &NodeDb<'a>) -> Option<ExprPrime> {
        self.children(db).get(1).map(|e| ExprPrime(*e))
    }
}

syntax_node!(Term);

impl Term {
    /// Head of the term
    pub fn head<'a>(self, db: &NodeDb<'a>) -> Factor {
        Factor(self.children(db)[0])
    }

    /// Tail of the term
    pub fn tail<'a>(self, db: &NodeDb<'a>) -> Option<TermPrime> {
        self.children(db).get(1).map(|n| TermPrime(*n))
    }
}

syntax_node!(TermPrime);

impl TermPrime {
    /// The operation
    pub fn op<'a>(self, db: &NodeDb<'a>) -> RelationalOp {
        self.get_node(db).ty().clone().into()
    }

    /// Rhs
    pub fn rhs<'a>(self, db: &NodeDb<'a>) -> Factor {
        Factor(self.children(db)[0])
    }

    /// Optional tail
    pub fn tail<'a>(self, db: &NodeDb<'a>) -> Option<TermPrime> {
        self.children(db).get(1).map(|n| TermPrime(*n))
    }
}

syntax_node!(Factor);

impl Factor {
    /// The head of a factor
    pub fn head<'a>(self, db: &NodeDb<'a>) -> Product {
        Product(self.children(db)[0])
    }

    /// Optional tail
    pub fn tail<'a>(self, db: &NodeDb<'a>) -> Option<FactorPrime> {
        self.children(db).get(1).map(|n| FactorPrime(*n))
    }
}

syntax_node!(FactorPrime);

impl FactorPrime {
    /// The operation in the tail
    pub fn op<'a>(self, db: &NodeDb<'a>) -> AdditiveOp {
        self.get_node(db).ty().clone().into()
    }

    /// The rhs of the tail
    pub fn rhs<'a>(self, db: &NodeDb<'a>) -> Product {
        Product(self.children(db)[0])
    }

    /// Optional tail
    pub fn tail<'a>(self, db: &NodeDb<'a>) -> Option<FactorPrime> {
        self.children(db).get(1).map(|n| FactorPrime(*n))
    }
}

syntax_node!(Product);

impl Product {
    /// The head
    pub fn head<'a>(self, db: &NodeDb<'a>) -> Atom {
        Atom(self.children(db)[0])
    }

    /// Opt tail
    pub fn tail<'a>(self, db: &NodeDb<'a>) -> Option<ProductPrime> {
        self.children(db).get(1).map(|n| ProductPrime(*n))
    }
}

syntax_node!(ProductPrime);

impl ProductPrime {
    /// The operation
    pub fn op<'a>(self, db: &NodeDb<'a>) -> MultiplicativeOp {
        self.get_node(db).ty().clone().into()
    }

    /// The rhs of the product
    pub fn rhs<'a>(self, db: &NodeDb<'a>) -> Atom {
        Atom(self.children(db)[0])
    }

    /// Optional tail
    pub fn tail<'a>(self, db: &NodeDb<'a>) -> Option<ProductPrime> {
        self.children(db).get(1).map(|n| ProductPrime(*n))
    }
}

syntax_node!(Atom);

/// Possible types of an atom
pub enum AtomType {
    /// Literal, function call or bracketed expr
    Unit(Unit),

    /// The not of an atom
    Not(Atom),
}

impl Atom {
    /// Convert an atom to a more specific type
    pub fn downcast<'a>(self, db: &NodeDb<'a>) -> AtomType {
        let n = self.children(db)[0];
        match self.get_node(db).ty() {
            NodeType::Not => AtomType::Not(Atom(n)),
            NodeType::Unit => AtomType::Unit(Unit(n)),
            _ => panic!("Invalid atom type"),
        }
    }
}

syntax_node!(Unit);

impl Unit {
    /// Convert a unit into a more specific type
    pub fn downcast<'a>(self, db: &NodeDb<'a>) -> UnitType {
        match self.get_node(db).ty() {
            NodeType::Int(i) => UnitType::Int(*i),
            NodeType::Bool(b) => UnitType::Bool(*b),
            NodeType::Str => UnitType::Str(self.0),
            NodeType::Identifier => UnitType::Identifier(Identifier(self.children(db)[0])),
            NodeType::FunctionCall => UnitType::FunctionCall(FunctionCall(self.children(db)[0])),
            NodeType::BracketedExpr => UnitType::BracketedExpr(Expr(self.children(db)[0])),
            _ => panic!("Invalid atom type"),
        }
    }
}

/// Possible unit types
#[derive(Debug, Clone)]
#[allow(missing_docs)]
pub enum UnitType {
    Int(isize),
    Str(NodeId),
    Bool(bool),
    Identifier(Identifier),
    FunctionCall(FunctionCall),
    BracketedExpr(Expr),
}

syntax_node!(Identifier);

impl Identifier {
    /// Get the string id
    pub fn id<'a>(self, db: &NodeDb<'a>, text: &'a str) -> &'a str {
        use crate::range::Ranged;
        &text[self.get_node(db).range().clone()]
    }
}

#[allow(missing_docs)]
#[derive(Debug, Clone)]
pub enum AdditiveOp {
    Plus,
    Minus,
}

#[allow(missing_docs)]
#[derive(Debug, Clone)]
pub enum MultiplicativeOp {
    Times,
    Divide,
}

#[allow(missing_docs)]
#[derive(Debug, Clone)]
pub enum RelationalOp {
    LesserThan,
    GreaterThan,
    Equals,
    GreaterOrEquals,
    LesserOrEquals,
}

#[allow(missing_docs)]
#[derive(Debug, Clone)]
pub enum BooleanOp {
    And,
    Or,
}

/// Enum to convert any node into a typed one
#[allow(missing_docs)]
#[derive(Debug, Clone)]
pub enum SyntaxNode {
    Program(Program),
    Compound(Compound),
    Decl(Decl),
    PrintStat(PrintStat),
    Print(Print),
    Println(Println),
    Get(Get),
    While(While),
    If(If),
    Assign(Assign),
    FunctionDecl(FunctionDecl),
    FunctionDeclArgs(FunctionDeclArgs),
    FunctionCall(FunctionCall),
    FunctionCallArgs(FunctionCallArgs),
    Return(Return),
    Expr(Expr),
    ExprPrime(ExprPrime),
    Term(Term),
    TermPrime(TermPrime),
    Factor(Factor),
    FactorPrime(FactorPrime),
    Product(Product),
    ProductPrime(ProductPrime),
    Atom(Atom),
    Unit(Unit),
    Identifier(Identifier),
}

impl SyntaxNode {
    /// Convert a id to a syntax node
    pub fn from<'a>(input: NodeId, db: &NodeDb<'a>) -> Self {
        match db.get_node(input).expect("Invalid id").ty() {
            NodeType::Program => SyntaxNode::Program(Program(input)),
            NodeType::Compound => SyntaxNode::Compound(Compound(input)),
            NodeType::Decl => SyntaxNode::Decl(Decl(input)),
            NodeType::PrintStat => SyntaxNode::PrintStat(PrintStat(input)),

            NodeType::While => SyntaxNode::While(While(input)),
            NodeType::If => SyntaxNode::If(If(input)),
            NodeType::Assign => SyntaxNode::Assign(Assign(input)),
            NodeType::FunctionCall => SyntaxNode::FunctionCall(FunctionCall(input)),
            NodeType::FunctionDecl => SyntaxNode::FunctionDecl(FunctionDecl(input)),
            NodeType::Return => SyntaxNode::Return(Return(input)),
            NodeType::Print => SyntaxNode::Print(Print(input)),
            NodeType::Println => SyntaxNode::Println(Println(input)),
            NodeType::Get => SyntaxNode::Get(Get(input)),
            NodeType::FunctionDeclArgs => SyntaxNode::FunctionDeclArgs(FunctionDeclArgs(input)),
            NodeType::FunctionCallArgs => SyntaxNode::FunctionCallArgs(FunctionCallArgs(input)),

            NodeType::Expr => SyntaxNode::Expr(Expr(input)),
            NodeType::And | NodeType::Or => SyntaxNode::ExprPrime(ExprPrime(input)),

            NodeType::Term => SyntaxNode::Term(Term(input)),
            NodeType::LesserThan
            | NodeType::GreaterThan
            | NodeType::Equals
            | NodeType::GreaterOrEquals
            | NodeType::LesserOrEquals => SyntaxNode::TermPrime(TermPrime(input)),

            NodeType::Factor => SyntaxNode::Factor(Factor(input)),
            NodeType::Plus | NodeType::Minus => SyntaxNode::FactorPrime(FactorPrime(input)),

            NodeType::Product => SyntaxNode::Product(Product(input)),
            NodeType::Times | NodeType::Divide => SyntaxNode::ProductPrime(ProductPrime(input)),

            NodeType::Unit | NodeType::Not => SyntaxNode::Atom(Atom(input)),

            NodeType::Int(_) | NodeType::Str | NodeType::Bool(_) | NodeType::BracketedExpr => {
                SyntaxNode::Unit(Unit(input))
            }

            NodeType::Identifier => SyntaxNode::Identifier(Identifier(input)),

            NodeType::Error(_) => panic!("Called on error node"),
        }
    }
}
