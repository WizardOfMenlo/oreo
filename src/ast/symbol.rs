//! Utilities for figuring out symbols and scopes

use super::node_db::NodeDb;
use super::syntax::*;
use std::collections::HashMap;

/// An id for an identifier
#[derive(Debug, Clone, Copy, Hash, Eq, PartialEq)]
pub struct IdentId(usize);

/// A id for a scope
#[derive(Debug, Clone, Copy, Hash, Eq, PartialEq)]
pub struct ScopeId(usize);

/// A full symbol table
#[derive(Debug, Default)]
pub struct SymbolTable<'a> {
    scopes: HashMap<ScopeId, Scope<'a>>,
}

/// A scope
#[derive(Debug, Clone)]
pub struct Scope<'a> {
    parent: ScopeId,
    variables: PartialSymbolTable<'a>,
}

/// A symbol table for a scope
#[derive(Debug, Default, Clone)]
pub struct PartialSymbolTable<'a> {
    vars: Vec<VariableRecord<'a>>,
}

/// A possible declaration
#[derive(Debug, Clone)]
pub enum DeclarationContext {
    /// var x := 0
    Decl(Decl),
    /// procedure f(var f) ...
    FunctionArg(FunctionDeclArgs),
}

/// A record for a variable (i.e. declaration and name)
#[derive(Debug, Clone)]
pub struct VariableRecord<'a> {
    id: IdentId,
    text: &'a str,
    decl: DeclarationContext,
}

/// Build the symbol table from a node
#[derive(Debug)]
pub struct SymbolTableBuilder<'a, 'b> {
    current_id: IdentId,
    current_scope: ScopeId,
    // Used to generate new ids
    scope_count: usize,
    symb: SymbolTable<'a>,
    input: &'a str,
    db: &'b NodeDb<'a>,
}

impl<'a, 'b> SymbolTableBuilder<'a, 'b> {
    /// Starts building the symbol table
    pub fn new(input: &'a str, db: &'b NodeDb<'a>) -> Self {
        SymbolTableBuilder {
            current_id: IdentId(0),
            current_scope: ScopeId(0),
            scope_count: 0,
            symb: SymbolTable::default(),
            input,
            db,
        }
    }

    /// Build the full table
    pub fn build(mut self, program: Program) -> SymbolTable<'a> {
        self.compound(program.compound(self.db));

        self.symb
    }

    fn get_current_scope(&mut self) -> &mut Scope<'a> {
        self.symb
            .scopes
            .get_mut(&self.current_scope)
            .expect("Invalid scope")
    }

    fn enter_new_scope(&mut self) {
        dbg!(&self.symb);
        let new_scope = Scope {
            parent: self.current_scope,
            variables: PartialSymbolTable::default(),
        };

        self.scope_count += 1;
        self.current_scope = ScopeId(self.scope_count);
        self.symb.scopes.insert(self.current_scope, new_scope);

        dbg!(&self.symb);
    }

    fn exit_scope(&mut self) {
        self.current_scope = self.get_current_scope().parent;
    }

    fn add_var(&mut self, id: Identifier, decl: DeclarationContext) {
        let curr_id = self.current_id;
        let id_str = id.id(self.db, self.input);
        self.get_current_scope()
            .variables
            .vars
            .push(VariableRecord {
                id: curr_id,
                text: id_str,
                decl,
            });

        self.current_id = IdentId(self.current_id.0 + 1);
    }

    fn compound(&mut self, compound: Compound) {
        self.enter_new_scope();
        for statement in compound.statements(self.db) {
            match statement.downcast(self.db) {
                StatementType::Decl(d) => {
                    self.add_var(d.id(self.db), DeclarationContext::Decl(d));
                }
                StatementType::FunctionDecl(f) => {
                    self.enter_new_scope();
                    for arg in f.args(self.db) {
                        self.add_var(arg.id(self.db), DeclarationContext::FunctionArg(arg));
                    }
                    self.compound(f.compound(self.db));
                    self.exit_scope();
                }
                StatementType::If(i) => {
                    self.compound(i.if_branch(self.db));
                    if let Some(c) = i.else_branch(self.db) {
                        self.compound(c);
                    }
                }
                StatementType::While(w) => {
                    self.compound(w.compound(self.db));
                }
                _ => continue,
            }
        }
        self.exit_scope();
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_debug_snapshot;
    use std::pin::Pin;

    fn db_from_str(input: &str) -> Pin<Box<NodeDb>> {
        use crate::lexer::lexicalize;
        use crate::lexer::scanner::scan;
        use crate::parser::parse;

        let node = parse(lexicalize(scan(input)));
        NodeDb::new(node)
    }

    // We use this to have a determined order
    fn determinize<'a>(table: SymbolTable<'a>) -> Vec<(ScopeId, Scope<'a>)> {
        let mut v: Vec<_> = table.scopes.into_iter().collect();
        v.sort_by(|a, b| (a.0).0.cmp(&(b.0).0));
        v
    }

    #[test]
    fn test_simple() {
        let input = "program x begin var x := 10; end";
        let db = db_from_str(input);
        let table = SymbolTableBuilder::new(input, &db).build(Program::new(db.start_id()));
        assert_debug_snapshot!(determinize(table));
    }

    #[test]
    fn test_nested() {
        let input = r#"program id begin
                   if ( true ) then begin var n := 5;  end
       			   else begin var k := 46; end; end"#;
        let db = db_from_str(input);
        let table = SymbolTableBuilder::new(input, &db).build(Program::new(db.start_id()));
        assert_debug_snapshot!(determinize(table));
    }

    #[test]
    fn test_fun_decl() {
        let input = r#"
        program id begin
            procedure f(var x, var y) begin
                var z;
                var k;
            end
       	end"#;
        let db = db_from_str(input);
        let table = SymbolTableBuilder::new(input, &db).build(Program::new(db.start_id()));
        assert_debug_snapshot!(determinize(table));
    }
}
