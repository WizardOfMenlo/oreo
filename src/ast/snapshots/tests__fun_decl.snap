---
source: src/ast/symbol.rs
expression: determinize(table)
---
[
    (
        ScopeId(
            1,
        ),
        Scope {
            parent: ScopeId(
                0,
            ),
            variables: PartialSymbolTable {
                vars: [],
            },
        },
    ),
    (
        ScopeId(
            2,
        ),
        Scope {
            parent: ScopeId(
                1,
            ),
            variables: PartialSymbolTable {
                vars: [
                    VariableRecord {
                        id: IdentId(
                            0,
                        ),
                        text: "x",
                        decl: NodeId(
                            3,
                        ),
                    },
                    VariableRecord {
                        id: IdentId(
                            1,
                        ),
                        text: "y",
                        decl: NodeId(
                            5,
                        ),
                    },
                ],
            },
        },
    ),
    (
        ScopeId(
            3,
        ),
        Scope {
            parent: ScopeId(
                2,
            ),
            variables: PartialSymbolTable {
                vars: [
                    VariableRecord {
                        id: IdentId(
                            2,
                        ),
                        text: "z",
                        decl: NodeId(
                            7,
                        ),
                    },
                    VariableRecord {
                        id: IdentId(
                            3,
                        ),
                        text: "k",
                        decl: NodeId(
                            9,
                        ),
                    },
                ],
            },
        },
    ),
]
